File: sdk/csharp/driversdk/DriverSDK/Plugin/GrpcDriver.cs
using Grpc.Core;
using Proto;

namespace DriverSDK.Plugin
{
	public class GrpcDriver: Driver.DriverBase
	{
        private IDriver _driver; 
        private IReport? _report; 

        public GrpcDriver(IDriver driver)
        {
            _driver = driver; 
        }

        public override Task<ResponseResult> GetDriverInfo(RequestArgs request, ServerCallContext context)
        {
            return Task.FromResult(new ResponseResult { Data = _driver.GetDriverInfo(request.Request) });
        }

        public override Task<ResponseResult> SetConfig(RequestArgs request, ServerCallContext context)
        {
            return Task.FromResult(new ResponseResult { Data = _driver.SetConfig(request.Request) });
        }

        public override Task<ResponseResult> Setup(RequestArgs request, ServerCallContext context)
        {
            _report = new GrpcReport("0.0.0.0", (int)request.Brokerid);
            return Task.FromResult(new ResponseResult { Data = _driver.Setup(request.Request, _report) });
        }

        public override Task<ResponseResult> Start(RequestArgs request, ServerCallContext context)
        {
            return Task.FromResult(new ResponseResult { Data = _driver.Start(request.Request) });
        }

        public override Task<ResponseResult> Restart(RequestArgs request, ServerCallContext context)
        {
            return Task.FromResult(new ResponseResult { Data = _driver.Restart(request.Request) });
        }

        public override Task<ResponseResult> Stop(RequestArgs request, ServerCallContext context)
        {
            return Task.FromResult(new ResponseResult { Data = _driver.Stop(request.Request) });
        }

        public override Task<ResponseResult> Get(RequestArgs request, ServerCallContext context)
        {
            return Task.FromResult(new ResponseResult { Data = _driver.Get(request.Request) });
        }

        public override Task<ResponseResult> Set(RequestArgs request, ServerCallContext context)
        {
            return Task.FromResult(new ResponseResult { Data = _driver.Set(request.Request) });
        }
    }
}

File: sdk/cxx/driversdk/src/plugin/grpcdriver.cpp
#include "grpcdriver.h"
#include "grpcreport.h"

GrpcDriver::GrpcDriver(IDriver &idriver) : driver(idriver) {}

grpc::Status GrpcDriver::GetDriverInfo(grpc::ServerContext *context, const proto::RequestArgs *request,
                                       proto::ResponseResult *response) {
    std::string data = driver.getDriverInfo(request->request());
    response->set_data(data);
    return grpc::Status::OK;
}

grpc::Status GrpcDriver::SetConfig(grpc::ServerContext *context, const proto::RequestArgs *request,
                                   proto::ResponseResult *response) {
    std::string data = driver.setConfig(request->request());
    response->set_data(data);
    return grpc::Status::OK;
}

grpc::Status
GrpcDriver::Setup(grpc::ServerContext *context, const proto::RequestArgs *request, proto::ResponseResult *response) {
    std::string port = std::to_string(request->brokerid());
    std::string addr = "0.0.0.0:" + port;
    GrpcReport report(addr);

    std::string data = driver.setup(request->request(), &report);
    response->set_data(data);
    return grpc::Status::OK;
}

grpc::Status
GrpcDriver::Start(grpc::ServerContext *context, const proto::RequestArgs *request, proto::ResponseResult *response) {
    std::string data = driver.start(request->request());
    response->set_data(data);
    return grpc::Status::OK;
}

grpc::Status
GrpcDriver::Restart(grpc::ServerContext *context, const proto::RequestArgs *request, proto::ResponseResult *response) {
    std::string data = driver.restart(request->request());
    response->set_data(data);
    return grpc::Status::OK;
}

grpc::Status
GrpcDriver::Stop(grpc::ServerContext *context, const proto::RequestArgs *request, proto::ResponseResult *response) {
    std::string data = driver.stop(request->request());
    response->set_data(data);
    return grpc::Status::OK;
}

grpc::Status
GrpcDriver::Get(grpc::ServerContext *context, const proto::RequestArgs *request, proto::ResponseResult *response) {
    std::string data = driver.get(request->request());
    response->set_data(data);
    return grpc::Status::OK;
}

grpc::Status
GrpcDriver::Set(grpc::ServerContext *context, const proto::RequestArgs *request, proto::ResponseResult *response) {
    std::string data = driver.set(request->request());
    response->set_data(data);
    return grpc::Status::OK;
}

